### DISCUSSION TITLE
ttgo t call v1 3 esp32 sim800l with modbus RS-485 Read holding regsiters

### DISCUSSION THREAD

**Discussion by CH-Fawez** [eModbus/eModbus#291] [eModbus applications]
Hello everyone,

I am currently working on a project where I need to use the TTGO T-Call V1.3 ESP32 SIM800L with Modbus to send a request with CRC check. Specifically, I need to send the following request: (0x01, 0x03, 0x02, 0xA9, 0x00, 0x21).

I have been researching and trying to figure out how to do this for a while now, but I am still having trouble getting it to work. I have come across the emodbus library, which seems to be a good option for implementing Modbus communication, but I am not sure how to use it for this specific request.

Could anyone provide me with some guidance on how to use the TTGO T-Call V1.3 ESP32 SIM800L with Modbus to send this request with CRC check using emodbus? Any help would be greatly appreciated.

Thank you in advance!

**Comment by Miq1** [comment]
There are some details still unclear to me. Are you intending to do the request from the ESP as a client on the RS485 RTU Modbus to some server, or are you trying to send a request via mobile network to the ESP acting as a server?

The former would be easy, while the latter could pose some problems that need to be solved in regards to the mobile connection.

**Reply by CH-Fawez** [reply]
yess i need to send request from my esp32 to a plc and receive the buffer of data then i need to send the data to server using the gprs module via tcp ip

**Reply by CH-Fawez** [reply]
could u help me please??

**Reply by Miq1** [reply]
Have a look at the RTU03example in the examples section - it should pretty much demonstrate what you are trying to achieve.
With the library you do not have to care about the CRC, it will be handled automatically.

**Reply by CH-Fawez** [reply]
i'm new in embedded systems thank you for your help <3 i'll keep you in touch

**Reply by CH-Fawez** [reply]
to send this request what i should change please ? (0x01, 0x03, 0x02, 0xA9, 0x00, 0x21).

**Reply by Miq1** [reply]
Modbus basics, I am afraid...

0x01 is the server ID of the requested server, 0x03 the function code for READ_HOLD_REGISTER, 0x02A9 the register address and 33 (0x21) the number of registers to be read.

In the example sketch you will find a line like ``err = addRequest(...`` where these parameters are set. Have a look at the comment lines before that line to tell which parameter is what data. Modify the line to your parameters and give it a try.

I suppose you have sought for information on how to connect to RS485 and have wired your ESP accordingly? 😉

**Reply by CH-Fawez** [reply]
yes i know how to connect to RS485  
in the sketch i didn't found line like ### **err = addRequest(...** 

```
**### // =================================================================================================
// eModbus: Copyright 2020 by Michael Harwerth, Bert Melis and the contributors to ModbusClient
//               MIT license - see license.md for details
// =================================================================================================
// Includes: <Arduino.h> for Serial etc., WiFi.h for WiFi support
#include <Arduino.h>
#include "HardwareSerial.h"

// Modbus server include
#include "ModbusServerRTU.h"

// Create a ModbusRTU server instance listening with 2000ms timeout
ModbusServerRTU MBserver(2000);

// FC03: worker do serve Modbus function code 0x03 (READ_HOLD_REGISTER)
ModbusMessage FC03(ModbusMessage request) {
  uint16_t address;           // requested register address
  uint16_t words;             // requested number of registers
  ModbusMessage response;     // response message to be sent back

  // get request values
  request.get(2, address);
  request.get(4, words);

  // Address and words valid? We assume 10 registers here for demo
  if (address && words && (address + words) <= 10) {
    // Looks okay. Set up message with serverID, FC and length of data
    response.add(request.getServerID(), request.getFunctionCode(), (uint8_t)(words * 2));
    // Fill response with requested data
    for (uint16_t i = address; i < address + words; ++i) {
      response.add(i);
    }
  } else {
    // No, either address or words are outside the limits. Set up error response.
    response.setError(request.getServerID(), request.getFunctionCode(), ILLEGAL_DATA_ADDRESS);
  }
  return response;
}

// Setup() - initialization happens here
void setup() {
// Init Serial monitor
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

// Init Serial2 connected to the RTU Modbus
// (Fill in your data here!)
  RTUutils::prepareHardwareSerial(Serial2);
  Serial2.begin(19200, SERIAL_8N1, GPIO_NUM_17, GPIO_NUM_16);

// Register served function code worker for server 1, FC 0x03
  MBserver.registerWorker(0x01, READ_HOLD_REGISTER, &FC03);

// Start ModbusRTU background task
  MBserver.begin(Serial2);
}

// loop() - nothing done here today!
void loop() {
  delay(10000);
}**
```

**Reply by Miq1** [reply]
Aw, sorry! That was the server-side example. 😄 
Try ``RTU04example`` instead. There, line 91 reads
```
Error err = MB.addRequest((uint32_t)millis(), 1, READ_INPUT_REGISTER, FIRST_REGISTER, NUM_VALUES * 2);
```
You need to change ``READ_INPUT_REGISTER`` into ``READ_HOLD_REGISTER``, ``FIRST_REGISTER`` will have to be set to your 0x02A9 and ``NUM_VALUES *2`` replaced by 33.

On line 28 the data type of ``values`` needs to be altered to ``uint16_t``.

The output on line 104 should read
```
Serial.printf("   %04X: %5d\n", i + FIRST_REGISTER, values[i]);
```
